Now I will import the VCF into R, and do additional filtering there.
First, I will convert the VCF into gds and seq.gds formats.
```
#!/usr/bin/env Rscript

### libraries
        library(gdsfmt) 
        library(SNPRelate)
        library(data.table)
        library(ggplot2)
        library(foreach)
        library(lattice)
        library(tidyr)
        library(SeqArray)
        library(tidyverse)

### format data properly to work with SNPRelate
        ### subsetted data for PCA stuff
                vcf.fn <- "/mnt/spicy_3/Karen/201620172018FinalMapping/totalnewmapwMarch2018_Afiltsnps10bpindels_snps_filter_pass_lowGQmiss.vcf"
                snpgdsVCF2GDS(vcf.fn, "/mnt/spicy_3/Karen/201620172018FinalMapping/totalnewmapwMarch2018_Afiltsnps10bpindels_snps_filter_pass_lowGQmiss.gds", 
                                        method=c("biallelic.only"), snpfirstdim = FALSE)
        
                seqVCF2GDS(vcf.fn, "/mnt/spicy_3/Karen/201620172018FinalMapping/totalnewmapwMarch2018_Afiltsnps10bpindels_snps_filter_pass_lowGQmiss.seq.gds")
```
Now let's open the file and take an intial loook at the SNPs.
```
### libraries
        library(gdsfmt) 
        library(SNPRelate)
        library(data.table)
        library(ggplot2)
        library(foreach)
        library(lattice)
        library(tidyr)
        library(SeqArray)
        library(tidyverse)

		genofile <- seqOpen("/mnt/spicy_3/Karen/201620172018FinalMapping/totalnewmapwMarch2018_Afiltsnps10bpindels_snps_filter_pass_lowGQmiss.seq.gds")

		snps <- data.table(variant.ids = seqGetData(genofile, "variant.id"),
					chr = seqGetData(genofile, "chromosome"),
					pos = seqGetData(genofile, "position"),
					dp = seqGetData(genofile, "annotation/info/DP"))

    ggplot(data=snps, aes(x=dp)) + geom_histogram()
	  ggplot(data=snps, aes(x=log10(dp))) + geom_histogram()
```
![SNPsInitialReadDepthDistribution](SNPsInititalReadDepthDistributionwMarch2018.tiff)
![SNPsInitialReadDepthDistributionLog10](SNPsInititalReadDepthDistributionwMarch2018Log10.tiff)

Now let's filter out SNPs that occur in areas flagged as having too high or low read depth when mapping the D84A 10X Illumina short reads to the D84Aimages reference genome. Will also filter out SNPs on the edges of runs of Ns and at the ends of scaffolds.
```
	NsChrRD <- fread("/mnt/spicy_3/Karen/RefGenome/Dovetail/HiCnew/NsandDepthandChrEnd.sorted.500merged.bed")
	colnames(NsChrRD) <- c("chr", "start", "stop")
	NsChrRD$count <- c(1:26531)

	setkey(snps, chr, pos)
	initialsnps <- snps$variant.ids

	NsChrRDsnps <- foreach(i=NsChrRD$count, .combine="c")%do%{
		
	c=NsChrRD$chr[[i]]
	s=NsChrRD$start[[i]]
	p=NsChrRD$stop[[i]]
		
	temp <- snps[J(data.table(chr=c, pos=c(s:p), key="chr,pos")), nomatch=0]
	temp$variant.ids
		
	}

	save(NsChrRDsnps, file="NsChrRDsnpswMarch2018_20190418.Rdata")

	seqSetFilter(genofile, variant.id=NsChrRDsnps)
		
	NsChrRDsnpssnps <- data.table(variant.ids = seqGetData(genofile, "variant.id"),
		chr = seqGetData(genofile, "chromosome"),
		pos = seqGetData(genofile, "position"),
		dp = seqGetData(genofile, "annotation/info/DP"))

	ggplot(data=NsChrRDsnpssnps, aes(x=dp)) + geom_histogram()		
	ggplot(data=NsChrRDsnpssnps, aes(x=log10(dp))) + geom_histogram()		
		
	goodsnpsnotinNsChrRD <- setdiff(initialsnps, NsChrRDsnps)
	
	seqSetFilter(genofile, variant.id=goodsnpsnotinNsChrRD)
	
	goodsnpsnotinNsChrRDtable <- data.table(variant.ids = seqGetData(genofile, "variant.id"),
		chr = seqGetData(genofile, "chromosome"),
		pos = seqGetData(genofile, "position"),
		dp = seqGetData(genofile, "annotation/info/DP"))

	ggplot(data=goodsnpsnotinNsChrRDtable, aes(x=dp)) + geom_histogram()
	ggplot(data=goodsnpsnotinNsChrRDtable, aes(x=log10(dp))) + geom_histogram()

	save(goodsnpsnotinNsChrRDtable, file="goodsnpsnotinNsChrRDtablewMarch2018_20190418.Rdata")


```
485,437 SNPs were removed based on this filtering. 2,385,040 SNPs remain. Here is the updated distribution of read depths.
![SNPDistributionofReadDepthFilterforNsChrEnd](AfterNsandScaffEndFiltSNPsDistReadDepthwMarch2018.tiff)
![SNPDistributionofReadDepthFilterforNsChrEndLog10](AfterNsandScaffEndFiltSNPsDistReadDepthwMarch2018Log10.tiff)

Now lets filter out snps based on overall read depth. Removing SNPs that are in the tails (too low and too high) of the distribution.
```
### Overall read depth filtering on snps without removing Repeat Masker

	quantile(goodsnpsnotinNsChrRDtable$dp,probs=c(0.015,0.975))

	#1.5% 97.5% 
	#1924 12311 

	lowRDsnps <- goodsnpsnotinNsChrRDtable[dp < 1924]
	highRDsnps <- goodsnpsnotinNsChrRDtable[dp > 12311]
	lowRDsnps$RD <- c("low")
	highRDsnps$RD <- c("high")
	lowhighRDsnps <- rbind(lowRDsnps, highRDsnps)
		
	save(lowhighRDsnps, file="lowhighRDsnps_wMarch2018_20190919.Rdata")
		
	dpfiltsnps <- goodsnpsnotinNsChrRDtable[dp > 1923 & dp < 12312]
	dpfiltsnpsids <- dpfiltsnps$variant.ids
		
	save(dpfiltsnps, file="dpfiltsnps_wMarch2018_20190419.Rdata")
		
	ggplot(data=dpfiltsnps, aes(x=dp)) + geom_histogram()
	ggplot(data=dpfiltsnps, aes(x=log10(dp))) + geom_histogram()
	
```
This depth filtering resulted in a removal of 95,301 SNPs, such that 2,289,739 SNPs remain. Here is the distribution of read depth of the remaining SNPs.
![ReadDepthDistribtionAfterDepthFiltering](ReadDepthDistributionsAfterDepthFiltering.tiff)

Now let's filter SNPs that have high or low read depth for specific superclones.

```
### Superclone specific depth filtering

	seqSetFilter(genofile, variant.id=dpfiltsnpsids)

	superclones <- fread("/mnt/spicy_3/Karen/Sp2017/supercloneassignment20180816.csv")
	colnames(superclones) <- c("clone", "pond", "year", "season", "SC094", "Plate", "SCnew", "SCnewer", "notes")
	superclones$clone <- ifelse(superclones$clone=="May_2017_DBunk_549", "May_2017_DBunk_549B", superclones$clone)

	Aclones <- superclones$clone[superclones$SC094=="A"]
	Bclones <- superclones$clone[superclones$SC094=="B"]
	Dclones <- superclones$clone[superclones$SC094=="D"]
	Hclones <- superclones$clone[superclones$SC094=="H"]
	Iclones <- superclones$clone[superclones$SC094=="I"]
	Lclones <- superclones$clone[superclones$SC094=="L"]
	Mclones <- superclones$clone[superclones$SC094=="M"]
		
	supercloneinput <- c("A", "B", "D", "H", "I", "L", "M")

	computingoutofdpsnps <- foreach(x=1:length(supercloneinput), .combine="rbind")%do%{
	
	tempgroup <- supercloneinput[x]
	tempgroupids <- superclones$clone[superclones$SC094==tempgroup]
	
	seqSetFilter(genofile, sample.id=tempgroupids)

### per-strain
    ### calculate depth
            
        dp <- t(seqGetData(genofile, "annotation/format/DP")$data)
        dp <- as.data.table(dp)
        dp[,variant.ids:=dpfiltsnpsids]
	
		sampleids <- as.data.table(seqGetData(genofile, "sample.id"))
		variantidhead <- as.data.table("variant.ids")
		totalids <- rbind(sampleids, variantidhead)
		colnames(dp) <- c(totalids$V1)
	
	    setkey(dp, variant.ids)
	    setkey(dpfiltsnps, variant.ids)
	    
	    m <- merge(dp, dpfiltsnps)

		readdepthrow <- melt(m, measure.vars=sampleids, variable.name="clone", value.name="dp")

		readdepthrow.ag <- readdepthrow[,list(totrd=as.double(sum(dp1, na.rm=TRUE))),
					 list(variant.ids)]
					 
		setkey(readdepthrow.ag, variant.ids)
		setkey(dpfiltsnps, variant.ids)
		mreaddepthrow.ag <- merge(readdepthrow.ag, dpfiltsnps)

		ggplot(data=mreaddepthrow.ag, aes(x=totrd)) + geom_histogram()
					 
		tempquant <- as.data.table(quantile(mreaddepthrow.ag$totrd,probs=c(0.025,0.975)))
		
 		
  		mreaddepthrow.aghigh <- mreaddepthrow.ag[totrd > tempquant$V1[[2]]]
  		mreaddepthrow.aghigh$highlow <- c("high")
   		mreaddepthrow.aglow <- mreaddepthrow.ag[totrd < tempquant$V1[[1]]]
   		mreaddepthrow.aglow$highlow <- c("low")
		mreaddepthrow.aglownottouse <- rbind(mreaddepthrow.aghigh, mreaddepthrow.aglow)
		mreaddepthrow.aglownottouse$clone <- tempgroup
		mreaddepthrow.aglownottouse$numind <- length(tempgroupids)
		mreaddepthrow.aglownottouse

	}

	save(computingoutofdpsnps, file="computingoutofdpsnps_wMarch2018_20190419.Rdata")

	computingoutofdpsnpsids <- data.table(variant.ids=computingoutofdpsnps$variant.ids)
	computingoutofdpsnpsidsuniq <- unique(computingoutofdpsnpsids)

	totalnotusescdepth <- computingoutofdpsnpsidsuniq$variant.ids

	dpfiltandscfilt <- setdiff(dpfiltsnpsids, totalnotusescdepth)

	seqSetFilter(genofile, variant.id=dpfiltandscfilt)

	snpsdpfiltscfilt <- data.table(variant.ids = seqGetData(genofile, "variant.id"),
		chr = seqGetData(genofile, "chromosome"),
		pos = seqGetData(genofile, "position"),
		dp = seqGetData(genofile, "annotation/info/DP"))

	save(dpfiltandscfilt, file="dpfiltandscfilt_wMarch2018_20190419.Rdata")

	ggplot(data=snpsdpfiltscfilt, aes(x=dp)) + geom_histogram()
```
This superclone specific depth filtering resulted in 327671 SNPs being removed, leaving 1,962,068 SNPs. Here is the read depth distribution of the remaining SNPs.
![ReadDepthDistribtionAfterSCDepthFiltering](ReadDepthAfterSCDepthFiltering.tiff)

Now lets remove trialleleic SNPs.

```
      	library(gdsfmt) 
        library(SNPRelate)
        library(data.table)
        library(ggplot2)
        library(foreach)
        library(lattice)
        library(tidyr)
        library(SeqArray)
        library(tidyverse)

	genofile <- seqOpen("/mnt/spicy_3/Karen/201620172018FinalMapping/totalnewmapwMarch2018_Afiltsnps10bpindels_snps_filter_pass_lowGQmiss.seq.gds")

	load(file="dpfiltandscfilt_wMarch2018_20190419.Rdata")
	
	seqSetFilter(genofile, variant.id=dpfiltandscfilt) #1,309,634
	
	tri <- (seqGetData(genofile, "$num_allele"))		
	tri <- as.data.table(tri)
	tri$variant.ids <- seqGetData(genofile, "variant.id")
	tri$diallelic <- ifelse(tri$tri=="2", 1, 0)
						
	updatesnpstouse <- tri$variant.ids[tri$tri=="2"]
			
	save(updatesnpstouse, file="updatesnpstouse_depthfiltandnotrialleleic_20190419.Rdata")
			
	seqSetFilter(genofile, variant.id=updatesnpstouse)
			
#	seqGDS2VCF(genofile, "updatesnpstouse_depthfiltandnotrialleleic_20190419.vcf")

```
Filtering for triallelic SNPs removed 30,819 SNPs. 1,931,249 SNPs remain.

Now let's try to figure out what Samples and SNPs should be dropped to minimize missing data.
```
	seqSetFilter(genofile, variant.id=updatesnpstouse)

	snpsG <- data.table(variant.ids = seqGetData(genofile, "variant.id"),
		chr = seqGetData(genofile, "chromosome"),
		pos = seqGetData(genofile, "position"),
		dp = seqGetData(genofile, "annotation/info/DP"))

#Pull out genotypes

	het <- t(seqGetData(genofile, "$dosage"))
	het <- as.data.table(het)
			
	colnames(het) <- c(seqGetData(genofile, "sample.id"))
	het$variant.ids <- seqGetData(genofile, "variant.id")
			
	setkey(het, variant.ids)
	setkey(snpsG, variant.ids)
			
	mhet <- merge(snpsG, het)

	save(mhet, file="mhet_wMarch2018_20190419")

# Set all genotypes to 1

	mhet[mhet == 0] <- 1
	mhet[mhet == 2] <- 1
	
	save(mhet, file="mhetgenosas1_20190422.Rdata")

# Sum genotypes across individuals, to get number of genotype sites per individual

	genocounts <- as.data.table(colSums(Filter(is.numeric, mhet), na.rm=TRUE))	

	genocounts <- genocounts[4:428]

	sample.ids <- seqGetData(genofile, "sample.id")

	genocounts$sample.ids <- sample.ids
	
	colnames(genocounts) <- c("numsitesgenotype", "sample.ids")
	
	setkey(genocounts, numsitesgenotype)

	save(genocounts, file="genocounts_20190422.Rdata")

# Let's look at the distribution of number of genotype sites across individuals

	ggplot(data=genocounts, aes(x=numsitesgenotype)) + geom_histogram(binwidth=10000)

	sampsbelow1375000 <- genocounts$sample.ids[genocounts$numsitesgenotype<1375000]
	sampsabove1375000 <- genocounts$sample.ids[genocounts$numsitesgenotype>1375000]

# Removing all individuals with sites genotyped below 1,375,000

	mhet[, (sampsbelow1375000) := NULL]

# Having done this, let's look at the distribution of number of individuals genotypes per SNP.

	mhetlong <- melt(mhet, measure.vars=sampsabove1375000, variable.name="clone", value.name="dosage")

	mhetlong.ag <- mhetlong[,list(numgeno = sum(dosage, na.rm=TRUE)), list(variant.ids) ]
	
	save(mhetlong.ag, file="mhetlong.ag_20190422.Rdata")



	#save(mhetlong, file="mhetlong_20190422.Rdata")
	
	#m[is.na(PA42),PA42:=FALSE]

